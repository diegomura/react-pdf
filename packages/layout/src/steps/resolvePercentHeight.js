import { isNil, matchPercent } from '@react-pdf/fns';

/**
 * Transform percent height into fixed
 *
 * @param {String | number} height
 * @return {number} height
 */
const transformHeight = (pageArea, height) => {
  const match = matchPercent(height);
  return match ? match.percent * pageArea : height;
};

/**
 * Get page area (height minus paddings)
 *
 * @param {Object} page
 * @return {number} page area
 */
const getPageArea = page => {
  const pageHeight = page.style.height;
  const pagePaddingTop = page.style?.paddingTop || 0;
  const pagePaddingBottom = page.style?.paddingBottom || 0;

  return pageHeight - pagePaddingTop - pagePaddingBottom;
};

/**
 * Transform node percent height to fixed
 *
 * @param {Object} page
 * @param {Object} node
 * @return {Object} transformed node
 */
const resolveNodePercentHeight = (page, node) => {
  if (isNil(page.style?.height)) return node;
  if (isNil(node.style?.height)) return node;

  const pageArea = getPageArea(page);
  const height = transformHeight(pageArea, node.style.height);
  const style = Object.assign({}, node.style, { height });

  return Object.assign({}, node, { style });
};

/**
 * Transform page immediate children with percent height to fixed
 *
 * @param {Object} page
 * @return {Object} transformed page
 */
const resolvePagePercentHeight = page => {
  if (!page.children) return page;

  const resolveChild = child => resolveNodePercentHeight(page, child);
  const children = page.children.map(resolveChild);

  return Object.assign({}, page, { children });
};

/**
 * Transform all page immediate children with percent height to fixed.
 * This is needed for computing correct dimensions on pre-pagination layout.
 *
 * @param {Object} document root
 * @return {Object} transformed document root
 */
const resolvePercentHeight = root => {
  if (!root.children) return root;

  const children = root.children.map(resolvePagePercentHeight);

  return Object.assign({}, root, { children });
};

export default resolvePercentHeight;
