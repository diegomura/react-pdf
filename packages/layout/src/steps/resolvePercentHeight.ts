import { isNil, matchPercent } from '@react-pdf/fns';

import { SafeDocumentNode, SafeNode, SafePageNode } from '../types';

/**
 * Transform percent height into fixed
 *
 * @param height
 * @returns Height
 */
const transformHeight = (pageArea: number, height: number | string) => {
  const match = matchPercent(height);
  return match ? match.percent * pageArea : height;
};

/**
 * Get page area (height minus paddings)
 *
 * @param page
 * @returns Page area
 */
const getPageArea = (page: SafePageNode) => {
  const pageHeight = page.style.height as number;
  const pagePaddingTop = (page.style?.paddingTop || 0) as number;
  const pagePaddingBottom = (page.style?.paddingBottom || 0) as number;

  return pageHeight - pagePaddingTop - pagePaddingBottom;
};

/**
 * Transform node percent height to fixed
 *
 * @param page
 * @param node
 * @returns Transformed node
 */
const resolveNodePercentHeight = (
  page: SafePageNode,
  node: SafeNode,
): SafeNode => {
  if (isNil(page.style?.height)) return node;
  if (isNil(node.style?.height)) return node;

  const pageArea = getPageArea(page);
  const height = transformHeight(pageArea, node.style.height);
  const style = Object.assign({}, node.style, { height });

  return Object.assign({}, node, { style });
};

/**
 * Transform page immediate children with percent height to fixed
 *
 * @param page
 * @returns Transformed page
 */
const resolvePagePercentHeight = (page: SafePageNode) => {
  if (!page.children) return page;

  const resolveChild = (child: SafeNode) =>
    resolveNodePercentHeight(page, child);

  const children = page.children.map(resolveChild);

  return Object.assign({}, page, { children });
};

/**
 * Transform all page immediate children with percent height to fixed.
 * This is needed for computing correct dimensions on pre-pagination layout.
 *
 * @param root - Document root
 * @returns Transformed document root
 */
const resolvePercentHeight = (root: SafeDocumentNode) => {
  if (!root.children) return root;

  const children = root.children.map(resolvePagePercentHeight);

  return Object.assign({}, root, { children });
};

export default resolvePercentHeight;
