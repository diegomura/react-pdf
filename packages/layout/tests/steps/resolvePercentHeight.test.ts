import { describe, expect, test } from 'vitest';

import resolvePercentHeight from '../../src/steps/resolvePercentHeight';

describe('layout resolvePercentHeight', () => {
  test('Should keep empty document untouched', () => {
    const result = resolvePercentHeight({
      type: 'DOCUMENT',
      props: {},
      children: [],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should keep empty page untouched', () => {
    const result = resolvePercentHeight({
      type: 'DOCUMENT',
      props: {},
      children: [{ type: 'PAGE', props: {}, style: {}, children: [] }],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should not resolve children if page dont have height', () => {
    const result = resolvePercentHeight({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: {},
          children: [
            { type: 'VIEW', props: {}, style: { width: '60%', height: '80%' } },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve children percent dimensions if page have height', () => {
    const result = resolvePercentHeight({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: { height: 1000 },
          children: [{ type: 'VIEW', props: {}, style: { height: '80%' } }],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });
});
