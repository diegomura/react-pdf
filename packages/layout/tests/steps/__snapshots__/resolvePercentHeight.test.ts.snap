// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layout resolvePercentHeight > Should keep empty document untouched 1`] = `
{
  "children": [],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePercentHeight > Should keep empty page untouched 1`] = `
{
  "children": [
    {
      "children": [],
      "props": {},
      "style": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePercentHeight > Should not resolve children if page dont have height 1`] = `
{
  "children": [
    {
      "children": [
        {
          "props": {},
          "style": {
            "height": "80%",
            "width": "60%",
          },
          "type": "VIEW",
        },
      ],
      "props": {},
      "style": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolvePercentHeight > Should resolve children percent dimensions if page have height 1`] = `
{
  "children": [
    {
      "children": [
        {
          "props": {},
          "style": {
            "height": 800,
          },
          "type": "VIEW",
        },
      ],
      "props": {},
      "style": {
        "height": 1000,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;
